Use std::bit_cast as c10::bit_cast if available

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141035
* #140994
* #140720
* #140567
* #140566
* #140565
* #140564

Make what we're doing as obvious as possible to the compiler.

Differential Revision: [D66108811](https://our.internmc.facebook.com/intern/diff/D66108811/)

swolchok committed:
Use std::bit_cast as c10::bit_cast if available

Make what we're doing as obvious as possible to the compiler.

Differential Revision: [D66108811](https://our.internmc.facebook.com/intern/diff/D66108811/)

[ghstack-poisoned]

swolchok committed:
Update on "Use std::bit_cast as c10::bit_cast if available"


Make what we're doing as obvious as possible to the compiler.

Differential Revision: [D66108811](https://our.internmc.facebook.com/intern/diff/D66108811/)

[ghstack-poisoned]

Skylion007 commented on c10/util/bit_cast.h:
```diff
@@ -3,8 +3,18 @@
 #include <cstring>
 #include <type_traits>
 
+#if __has_include(<bit>) && (__cplusplus >= 202002L || (defined(__cpp_lib_bit_cast) && __cpp_lib_bit_cast >= 201806L))
```
```suggestion
#ifdef __cpp_lib_bit_cast
```

Skylion007 commented on c10/util/bit_cast.h:
```diff
@@ -3,8 +3,18 @@
+#include <bit>
+#define C10_HAVE_STD_BIT_CAST 1
+#else
+#define C10_HAVE_STD_BIT_CAST 0
```
Similar logic to the prev PR. Does this macro serve any useful purpose. Main difference is whether bit_cast is constexpr I guess?

malfet replied:
Generally speaking, it would be good to move those `C10_HAVE_...` macros to say `c10/macros/Macros.h` or define new `Compiler.h` there

As defining those in various headers can lead to a potential duplication and than divergence

swolchok replied:
I just wanted to avoid duplicating the condition in this file.

> defining those in various headers

what other header would reasonably define or consume this? this is the bit_cast header.

I can update the diff to undefine its temporary macro

swolchok committed:
Update on "Use std::bit_cast as c10::bit_cast if available"


Make what we're doing as obvious as possible to the compiler.

Differential Revision: [D66108811](https://our.internmc.facebook.com/intern/diff/D66108811/)

[ghstack-poisoned]

swolchok committed:
Update on "Use std::bit_cast as c10::bit_cast if available"


Make what we're doing as obvious as possible to the compiler.

Differential Revision: [D66108811](https://our.internmc.facebook.com/intern/diff/D66108811/)

[ghstack-poisoned]

MatzeB commented:
Getting build errors, does this need `#include <bit`>?

MatzeB commented:
(Seems the #include is already there, but something is failing here...)

MatzeB commented on c10/util/bit_cast.h:
```diff
@@ -3,8 +3,19 @@
 #include <cstring>
 #include <type_traits>
 
```
This needs `#include <version>` before testing `__cpp_lib_xxx`.
